hadoop-crypto/src/main/java/com/palantir/crypto2/hadoop/EncryptedFileSystem.java

```diff
@@ -98,25 +100,39 @@ public boolean rename(Path src, Path dst) throws IOException {
         return renamed;
     }
 
+    private void tryRemoveKey(Path path) {
+        try {
+            keyStore.remove(path.toString());
+        } catch (Exception e) {
+            log.warn("Unable to remove KeyMaterial for file: {}", path);
+        }
+    }
+
     @Override
     public boolean delete(Path path, boolean recursive) throws IOException {
-        // Existence check allows for unused keys to be cleaned up
-        boolean exists = fs.exists(path);
-        boolean deleted = fs.delete(path, recursive);
+        if (recursive) {
+            return recursiveDelete(path);
+        }
+        return nonRecursiveDelete(path);
+    }
 
-        if (deleted || !exists) {
-            keyStore.remove(path.toString());
+    private boolean recursiveDelete(Path path) throws IOException {
+        //List all files under the path recursively and delete their key material
+        RemoteIterator<LocatedFileStatus> statuses = fs.listFiles(path, true);
+        while (statuses.hasNext()) {
+            LocatedFileStatus status = statuses.next();
+            if (status.isFile()) {
+                Path filePath = Path.getPathWithoutSchemeAndAuthority(status.getPath());
```

Comment:
@ellisjoe double check this - I believe this is fine since we're already wrapping the correct file system at this point. Existing tests need the scheme to be missing.